Add a `additional_mimes` parameter to the `add_image_size` function

adamsilverstein opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/56288

Related to: https://core.trac.wordpress.org/ticket/55443

<!-- insert a link to the WordPress Trac ticket here -->

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

adamsilverstein committed:
Add a boolean ‘output_mimes’ parameter to the add_image_size function (and adjust all uses)

adamsilverstein committed:
remove phpbf changes

felixarntz commented on src/wp-includes/media.php:
```diff
@@ -275,26 +275,29 @@ function image_downsize( $id, $size = 'medium' ) {
+ * @param bool       $output_mimes Whether to output secondary mimes for this image size. Default is false.
  */
-function add_image_size( $name, $width = 0, $height = 0, $crop = false ) {
+function add_image_size( $name, $width = 0, $height = 0, $crop = false, $output_mimes = false ) {
```
`$output_mimes` is a bit generic. This is only about additional / secondary MIME types. Maybe we should call it `$additional_mimes`?

adamsilverstein replied:
updated

felixarntz commented on src/wp-includes/media.php:
```diff
@@ -275,26 +275,29 @@ function image_downsize( $id, $size = 'medium' ) {
+ *                                 Array values must be in the format: array( x_crop_position, y_crop_position ) where:
+ *                                     - x_crop_position accepts: 'left', 'center', or 'right'.
+ *                                     - y_crop_position accepts: 'top', 'center', or 'bottom'.
+ * @param bool       $output_mimes Whether to output secondary mimes for this image size. Default is false.
```
Let's add the `Optional.` prefix to the description like on other optional parameters.

adamsilverstein replied:
updated

felixarntz commented on src/wp-includes/media.php:
```diff
@@ -343,9 +346,10 @@ function remove_image_size( $name ) {
  */
-function set_post_thumbnail_size( $width = 0, $height = 0, $crop = false ) {
-	add_image_size( 'post-thumbnail', $width, $height, $crop );
+function set_post_thumbnail_size( $width = 0, $height = 0, $crop = false, $output_mimes = true ) {
+	add_image_size( 'post-thumbnail', $width, $height, $crop, true );
```
It's a bit odd to have the parameter here, but also have `post-thumbnail` hard-coded in the list of core sizes above. Maybe we should omit this parameter on the function here for now and always set it to `true` inside? It's a bit odd to have it on `set_post_thumbnail_size()` when it has a different default than on `add_image_size()`.

adamsilverstein replied:
Ok, I can remove for now and we can consider later. main idea was to add a way to control the output when calling set_post_thumbnail_size but agree it feels odd.

adamsilverstein replied:
removed

felixarntz commented on src/wp-includes/media.php:
```diff
@@ -925,6 +929,10 @@ function wp_get_registered_image_subsizes() {
 
+		if ( isset( $additional_sizes[ $size_name ]['output_mimes'] ) ) {
+			$size_data['output_mimes'] = $additional_sizes[ $size_name ]['output_mimes'];
+		}
```
I'm not sure we need to include this here, since it's not really relevant when using the sizes. The `_wp_filter_image_sizes_additional_mime_type_support()` should be used to check for a list of sizes whether they have this flag set or not.

The problem of including this here is also that then it makes the data inconsistent, since core sizes now don't have the new flag set.

adamsilverstein replied:
removed

felixarntz commented on src/wp-includes/media.php:
```diff
@@ -5317,9 +5325,9 @@ function wp_media_personal_data_exporter( $email_address, $page = 1 ) {
+	add_image_size( '1536x1536', 1536, 1536, false, true );
 	// 2x large size.
-	add_image_size( '2048x2048', 2048, 2048 );
+	add_image_size( '2048x2048', 2048, 2048, false, true );
```
Related to my above comment on `post-thumbnail` size: We should also _only_ pass `true` there, but not make it a parameter on `set_post_thumbnail_size()`. This way it's consistent with the code here. All core sizes are opted in, only custom sizes are opted out; and if you want to opt out any core size, you'll have to use the filter.

adamsilverstein replied:
that makes sense, will do

adamsilverstein replied:
updated

felixarntz commented on tests/phpunit/tests/image/intermediateSize.php:
```diff
@@ -93,7 +93,7 @@ static function() {
 	 */
 	public function test_get_intermediate_sizes_by_name() {
-		add_image_size( 'test-size', 330, 220, true );
+		add_image_size( 'test-size', 330, 220, true, true );
```
Why pass the new parameter here (and throughout this file)? This seems pointless, we shouldn't need to pass it if it's optional and not relevant to the specific test.

felixarntz commented on tests/phpunit/tests/media.php:
```diff
@@ -1143,7 +1143,7 @@ public function test_add_image_size() {
 
 		$this->assertArrayNotHasKey( 'test-size', $_wp_additional_image_sizes );
-		add_image_size( 'test-size', 200, 600 );
+		add_image_size( 'test-size', 200, 600, false, true );
```
Same as above applies throughout this file.

adamsilverstein replied:
removed these changes

felixarntz commented on tests/phpunit/tests/media.php:
```diff
@@ -1169,7 +1169,7 @@ public function test_remove_image_size() {
-		add_image_size( 'test-size', 200, 600 );
+		add_image_size( 'test-size', 200, 600, false, true );
 		$this->assertTrue( has_image_size( 'test-size' ) );
 
```
Can you add some actual test coverage for the new parameter here? We should test that setting the parameter rightfully affects the additional registered sizes, and we should also make sure that it is taken into account when checking sizes via the `_wp_filter_image_sizes_additional_mime_type_support()` function.

adamsilverstein replied:
I'll add some coverage in `test_add_image_size`

adamsilverstein replied:
Added additional coverage specifically for the new `additional_mimes` parameter:

* Added coverage for both the expected data from `wp_get_additional_image_sizes` with `additional_mimes` true and false
* Added coverage that `_wp_filter_image_sizes_additional_mime_type_support()` also correctly returns the correct size where `additional_mimes` was set to true

felixarntz reviewed (commented):
@adamsilverstein This looks solid mostly. I left some feedback below.

Let's also add some test coverage, but I don't think it's worth updating all the other tests unnecessarily - the parameter is optional, so it seems pointless to set it in tests where it's not relevant.

adamsilverstein committed:
Remove test changes that aren’t required

adamsilverstein committed:
Update parameter name and doc block - use “additional_mimes”

adamsilverstein committed:
Skip setting output data on size

adamsilverstein committed:
Remove output mimes parameter from set_post_thumbnail_size

adamsilverstein committed:
revert changes on one more test file

adamsilverstein committed:
revert unrelated changes from trunk

adamsilverstein committed:
Merge branch 'trunk' into add/add-image-size-mime-support

adamsilverstein committed:
complete rename of output_mimes to additional_mimes

adamsilverstein committed:
add test coverage for add_image_size additional mimes parameter

adamsilverstein committed:
Add test for _wp_filter_image_sizes_additional_mime_type_support with add_image_size additional_mimes parameter

adamsilverstein commented:
@felixarntz Addressed your feedback here & ready for another review whenever you have a chance

felixarntz commented on src/wp-includes/media.php:
```diff
@@ -275,26 +275,29 @@ function image_downsize( $id, $size = 'medium' ) {
+ *                                     Array values must be in the format: array( x_crop_position, y_crop_position ) where:
+ *                                         - x_crop_position accepts: 'left', 'center', or 'right'.
+ *                                         - y_crop_position accepts: 'top', 'center', or 'bottom'.
+ * @param bool       $additional_mimes Optional. Whether to output secondary mimes for this image size. Default is false.
```
This parameter doc should be a lot more extensive. It should briefly explain that this should only be set to `false` if the image size is exclusively used for special cases which are not part of user-facing frontend content (e.g. exclusively for emails or Open Graph tags). We may even want to include that the default may eventually change.

adamsilverstein replied:
Updated in fc40a45371, please review.

felixarntz commented on src/wp-includes/media.php:
```diff
@@ -275,26 +275,29 @@ function image_downsize( $id, $size = 'medium' ) {
+ * @param bool       $additional_mimes Optional. Whether to output secondary mimes for this image size. Default is false.
  */
-function add_image_size( $name, $width = 0, $height = 0, $crop = false ) {
+function add_image_size( $name, $width = 0, $height = 0, $crop = false, $additional_mimes = false ) {
```
I thought we wanted to make the default here `null` so that it is easily clear whether the value has been set (as `true` or `false`) or not.

In addition, this also helps clarify that `false` is not the real default - i.e. the default is "dynamic" and can change in the future. The function parameter default will always be `null`, but then if it's not passed, the core implementation will override it to either `true` or `false` (e.g. `false` in 6.1, `true` at some point later).

Also: Do we want to trigger a `_doing_it_wrong()` if the parameter is not passed? Or should we hold off until later? Note that if we don't trigger a warning here now, it may delay the point in time when we can change the default.

adamsilverstein replied:
Changing the parameter to null and add a `doing_it_wrong` which I agree is important for notifying developers.

adamsilverstein replied:
Updated in 3a218e83ebbef9e36e4f5c18a0df91c8d50d04d8

felixarntz commented on src/wp-admin/includes/image.php:
```diff
@@ -660,6 +660,19 @@ function _wp_filter_image_sizes_additional_mime_type_support( $sizes, $attachmen
 		'post-thumbnail' => true,
```
Let's remove `post-thumbnail` from this list in favor of https://github.com/WordPress/wordpress-develop/pull/3028/files#diff-4d28b6652f0dab208c19aa45e1cf09f3c967060d8cc77502bfcd473a3216b612R351 which you've already added.

adamsilverstein replied:
👍🏼 removed in 8853869e95d08c92868a5398cfc47623534cc5d4

felixarntz commented on src/wp-admin/includes/image.php:
```diff
@@ -660,6 +660,19 @@ function _wp_filter_image_sizes_additional_mime_type_support( $sizes, $attachmen
 	);
 
+	// Include any sizes that were added with `add_image_size`.
+	$additional_image_sizes = wp_get_additional_image_sizes();
+	$additional_image_sizes = array_filter(
+		$additional_image_sizes,
+		function( $size ) {
+			return isset( $size['additional_mimes'] ) && $size['additional_mimes'];
+		}
+	);
+
+	foreach ( array_keys( $additional_image_sizes ) as $size_name ) {
+		$enabled_sizes[ $size_name ] = true;
+	}
```
A bit more elegant:
```suggestion
	$additional_image_sizes = array_map(
		'__return_true',
		array_filter(
			wp_get_additional_image_sizes(),
			function( $size ) {
				return isset( $size['additional_mimes'] ) && $size['additional_mimes'];
			}
		)
	);
	$enabled_sizes          = array_merge( $enabled_sizes, $additional_image_sizes );
```

felixarntz commented on tests/phpunit/tests/media.php:
```diff
@@ -1155,6 +1155,42 @@ public function test_add_image_size() {
+		add_image_size( 'test-size-with-additional-mimes', 200, 600, false, true );
+		add_image_size( 'test-size-without-additional-mimes', 200, 600, false, false );
+
+		$all_sizes      = wp_get_additional_image_sizes();
```
The variable name here is confusing, these aren't "all sizes", only the additional sizes. With that said, I do think passing all sizes here makes sense, so we should probably use `wp_get_registered_image_subsizes()` instead.

adamsilverstein replied:
👍🏼 updated in bcefc3a372

felixarntz commented on tests/phpunit/tests/media.php:
```diff
@@ -1155,6 +1155,42 @@ public function test_add_image_size() {
+		remove_image_size( 'test-size-with-additional-mimes' );
+		remove_image_size( 'test-size-without-additional-mimes' );
+
+		$this->assertArrayHasKey( 'test-size-with-additional-mimes', $filtered_sizes );
+		$this->assertArrayNotHasKey( 'test-size-without-additional-mimes', $filtered_sizes );
```
Instead of those specific assertions, I think it would be a bit more precise and comprehensive to build up the full expected array here and then ensure that all expected sizes are included. Something like
```php
$this->assertSame( $expected_size_names, array_keys( $filtered_sizes ) );
```

adamsilverstein replied:
👍🏼 Good idea, updated in 2474368056

felixarntz reviewed (commented):
@adamsilverstein Mostly looks good to me. I left some minor feedback.

My main point is actually to focus on improving the parameter documentation. The special behavior we intend for this parameter should be clearly explained.

adamsilverstein committed:
Update src/wp-admin/includes/image.php

Co-authored-by: Felix Arntz <felixarntz@users.noreply.github.com>

adamsilverstein committed:
Remove ‘post-thumbnail’ from enabled_sizes in _wp_filter_image_sizes_additional_mime_type_support (added via add_image_size)

adamsilverstein committed:
Use wp_get_registered_image_subsizes() for all_images in test__wp_filter_image_sizes_additional_mime_type_support_with_add_image_size

adamsilverstein committed:
Merge branch 'trunk' into add/add-image-size-mime-support

adamsilverstein committed:
Improve assertion in public function test__wp_filter_image_sizes_additional_mime_type_support_with_add_image_size

adamsilverstein committed:
Set default for parameter to null and warn developers with `_doing_it_wrong`

adamsilverstein committed:
Update doc block

adamsilverstein commented:
@felixarntz updated based on your feedback, ready for another review

adamsilverstein committed:
Add the additional_mimes parameter in all test calls to add_image_size

Avoid a “doing it wrong” message - “You should provide a value for the `additional_mimes` parameter.”

felixarntz commented on src/wp-includes/media.php:
```diff
@@ -275,26 +275,36 @@ function image_downsize( $id, $size = 'medium' ) {
+ *                                     Array values must be in the format: array( x_crop_position, y_crop_position ) where:
+ *                                         - x_crop_position accepts: 'left', 'center', or 'right'.
+ *                                         - y_crop_position accepts: 'top', 'center', or 'bottom'.
+ * @param bool       $additional_mimes Optional. Whether to output secondary mimes for this image size. Default is null
+ *                                     but may change to true in future versions. Sizes should be registered with false
+ *                                     only if they are not intended for use in a front-end context.
```
```suggestion
 * @param bool|null  $additional_mimes Optional. Whether to output secondary mimes for this image size. If not passed,
 *                                     WordPres core will determine whether true or false. Sizes should be registered
 *                                     with false only if they are not intended for use in a front-end context.
```

felixarntz commented on src/wp-includes/media.php:
```diff
@@ -275,26 +275,36 @@ function image_downsize( $id, $size = 'medium' ) {
  * Register a new image size.
  *
  * @since 2.9.0
+ * @since 6.1.0 Add the $output_mimes parameter.
```
```suggestion
 * @since 6.1.0 Add the $additional_mimes parameter.
```

felixarntz commented on src/wp-includes/media.php:
```diff
@@ -275,26 +275,36 @@ function image_downsize( $id, $size = 'medium' ) {
+function add_image_size( $name, $width = 0, $height = 0, $crop = false, $additional_mimes = null ) {
 	global $_wp_additional_image_sizes;
 
+	// Trigger a warning if the size was registered without providing an `additional_mimes` parameter.
+	if ( null === $additional_mimes ) {
+		_doing_it_wrong( __FUNCTION__, __( 'You should provide a value for the `additional_mimes` parameter.' ), '6.1.0' );
+	}
```
We also need to set an actual boolean value here if no value is passed. This is where we'll initially set it to `false`, but it could later change.

Also, please use real HTML instead of Markdown :)
```suggestion
	if ( null === $additional_mimes ) {
		_doing_it_wrong(
			__FUNCTION__,
			sprintf(
				/* translators: %s: additional_mimes variable */
				__( 'You should provide a boolean value for the %s parameter.' ),
				'<code>$additional_mimes</code>'
			),
			'6.1.0'
		);
		$additional_mimes = false;
	}
```

felixarntz reviewed (commented):
@adamsilverstein Just few final fixes to make here.

adamsilverstein committed:
Update src/wp-includes/media.php

Co-authored-by: Felix Arntz <felixarntz@users.noreply.github.com>

adamsilverstein committed:
Update src/wp-includes/media.php

Co-authored-by: Felix Arntz <felixarntz@users.noreply.github.com>

adamsilverstein committed:
Update src/wp-includes/media.php

Co-authored-by: Felix Arntz <felixarntz@users.noreply.github.com>

adamsilverstein committed:
Merge branch 'trunk' into add/add-image-size-mime-support

adamsilverstein committed:
Merge branch 'trunk' into add/add-image-size-mime-support

adamsilverstein commented:
> Also: Do we want to trigger a _doing_it_wrong() if the parameter is not passed? Or should we hold off until later? Note that if we don't trigger a warning here now, it may delay the point in time when we can change the default.

Reconsidering this: the additional error logging might be excessive for anyone who has error logging enabled - any existing image sizes need to get updated or they will throw a warning... I feel we should give developers one release cycle (at least) before adding this notice.  what do you think?

So my idea would be to add this in 6.1 with a Dev Note, but no warning. The in 6.2 we can add the `doing_it_wrong` notice; finally, in 6.3 we will switch the default from `null` to `true`.

adamsilverstein committed:
remove _doing_it_wrong warning (and related test changes)

adamsilverstein commented:
I removed the `doing_it_wrong` warning for now in [`12986a3` (#3028)](https://github.com/WordPress/wordpress-develop/pull/3028/commits/12986a321140af252b61ce8d6f2c71420e40f9df) and will mention that on the Trac ticket as well.

felixarntz commented:
@adamsilverstein Now that https://core.trac.wordpress.org/changeset/54097 has been committed, I suggest we refresh this PR against latest `trunk` and then adjust.

Most of the code here is still relevant as is, we just need to change the integration point which now is no longer in `_wp_filter_image_sizes_additional_mime_type_support()` but in `wp_default_image_output_mapping()`.
